enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,252 @@
+    - (TBD - waiting for clarification on this request.  Routes aren't associated with node types.)
+- Enable HSTS for specific subdomains 
+    - (TBD - dependency on wildcard DNS record/certificate, or just the `includeSubDomains`)
+- Disable HSTS per-route when it is enabled globally
+
+### Non-Goals
+
+- HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.  As
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP 
+response is received over insecure transport, clients must ignore any present STS headers.
+- The current implementation has a `preload` directive for the `Strict-Transport-Security`
+header.  This is not an RFC 6797 directive and therefore its implementation may vary by
+browser.  It can be specified, but support may vary by implementation.
+
+## Proposal
+
+Each goal is addressed in the subsections below.
+
+#### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an 
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is regarded as an HSTS host
+    - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+    - `max-age` decrements over time
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name
+- `preload`: if present, tells the client to include this host in its host preload list so that 
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797 
+and is therefore client implementation-dependent)
+
+HSTS is currently implemented as a pattern in the HAProxy template and applied to `edge` and 
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+````
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+````
+#### Enable HSTS for all routes globally
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment 
+variable called `ROUTER_DEFAULT_HSTS`, configured by the administrator in the Ingress Operator.  `ROUTER_DEFAULT_HSTS` is detected in the HAProxy configuration file processing, and can be overridden
```

Comment:
> add a new router environment  variable called `ROUTER_DEFAULT_HSTS`, configured by the administrator in the Ingress Operator


We need some sort of user-facing API so that a cluster administrator can tell the ingress operator to set the `ROUTER_DEFAULT_HSTS` environment variable for an ingress controller's deployment. In general, we don't want cluster administrators manually modifying router deployments by hand.


I think it would make sense to add an option for requiring HSTS when possible to the Ingress Controller API. This would allow a cluster administrator to trivially enable "global" HSTS for individual ingress controllers, and would allow for different HSTS settings for different ingress controllers. How does that sound to you?

Reply:
Yup, something along those lines id imagine.